// expo-audio/index.js
var EventEmitter = require('eventemitter3');

var permissionResult = { granted: true, status: 'granted', expires: 'never', canAskAgain: true };

function AudioStream(options) {
  EventEmitter.call(this);
  this.options = options;
  this._streaming = false;
}
AudioStream.prototype = Object.create(EventEmitter.prototype);
AudioStream.prototype.constructor = AudioStream;
AudioStream.prototype.start = function () {
  this._streaming = true;
  return Promise.resolve();
};
AudioStream.prototype.stop = function () {
  this._streaming = false;
};

function AudioRecorder(options) {
  EventEmitter.call(this);
  this._recording = false;
}
AudioRecorder.prototype = Object.create(EventEmitter.prototype);
AudioRecorder.prototype.constructor = AudioRecorder;
AudioRecorder.prototype.prepareToRecordAsync = function () { return Promise.resolve(); };
AudioRecorder.prototype.record = function () { this._recording = true; };
AudioRecorder.prototype.stop = function () { this._recording = false; return Promise.resolve(); };

function AudioPlayer(source, updateInterval, keepAudioSessionActive, preferredForwardBufferDuration) {
  EventEmitter.call(this);
  this._playing = false;
}
AudioPlayer.prototype = Object.create(EventEmitter.prototype);
AudioPlayer.prototype.constructor = AudioPlayer;
AudioPlayer.prototype.replace = function () {};
AudioPlayer.prototype.play = function () { this._playing = true; };
AudioPlayer.prototype.pause = function () { this._playing = false; };
AudioPlayer.prototype.stop = function () { this._playing = false; };

var AudioModule = {
  AudioRecorder: AudioRecorder,
  AudioPlayer: AudioPlayer,
  AudioStream: AudioStream,
};

module.exports = {
  AudioModule: AudioModule,
  requestRecordingPermissionsAsync: function () { return Promise.resolve(permissionResult); },
};
