// react-native/index.js
var NativeEventEmitter = function () {
  return { addListener: function () { return { remove: function () {} }; } };
};

var noop = function () {};

module.exports = {
  NativeModules: { YggstackModule: {
    addRemoteTCPMapping: function(port, addr) { return Promise.resolve(true); },
    removeRemoteTCPMapping: function(port, addr) { return Promise.resolve(true); },
    addListener: function(name) {},
    removeListeners: function(count) {},
  }, AudioTrackModule: {
    createCallChannel: function() { return Promise.resolve(true); },
    init: function(sampleRate) { return Promise.resolve(true); },
    write: function(b64) { return Promise.resolve(0); },
    stop: function() { return Promise.resolve(true); },
  } },
  NativeEventEmitter: NativeEventEmitter,
  useColorScheme: function () { return 'light'; },
  Appearance: {
    getColorScheme: function () { return 'light'; },
    addChangeListener: function () {},
  },
  BackHandler: { exitApp: noop },
  View: noop,
  Text: noop,
  ScrollView: noop,
  FlatList: noop,
  TextInput: noop,
  TouchableOpacity: noop,
  Pressable: noop,
  KeyboardAvoidingView: noop,
};
