// react-native-tcp-socket/index.js
var EventEmitter = require('eventemitter3');

var nextId = 0;
function genId() { return ++nextId; }

function MockSocket() {
  EventEmitter.call(this);
  this._id = genId();
  this.localAddress = '127.0.0.1';
  this.localPort = 7777;
  this.remoteAddress = '::1';
  this.remotePort = 54321;
  this.remoteFamily = 'IPv6';
  this._encoding = undefined;
  this._destroyed = false;
  this._pending = false;
  this._connecting = false;
  this.writableNeedDrain = false;
  this.readableHighWaterMark = 16384;
  this.writableHighWaterMark = 16384;
}
MockSocket.prototype = Object.create(EventEmitter.prototype);
MockSocket.prototype.constructor = MockSocket;
MockSocket.prototype.connect = function(options, callback) {
  if (callback) this.once('connect', callback);
  this._connecting = true;
  this._pending = true;
  var self = this;
  process.nextTick(function() {
    self._connecting = false;
    self._pending = false;
    if (typeof options === 'object' && options.host) {
      self.remoteAddress = options.host;
      self.remotePort = options.port || 7777;
    }
    self.emit('connect');
  });
  return this;
};
MockSocket.prototype.setEncoding = function(encoding) {
  this._encoding = encoding;
  return this;
};
MockSocket.prototype.write = function(data, encoding, cb) {
  if (typeof encoding === 'function') { cb = encoding; encoding = undefined; }
  if (typeof data === 'function') { cb = data; data = undefined; }
  if (typeof cb === 'function') process.nextTick(cb);
  this.emit('_write', data);
  // Auto-respond to SOCKS5 protocol
  if (data && data.length !== undefined && data.length >= 3 &&
      data[0] === 5 && data[1] === 1 && data[2] === 0) {
    var self = this;
    if (data.length === 3) {
      // SOCKS5 handshake: respond with auth success
      process.nextTick(function() { self.emit('data', new Uint8Array([5, 0])); });
    } else {
      // SOCKS5 CONNECT: respond with success
      process.nextTick(function() { self.emit('data', new Uint8Array([5, 0, 0, 1, 127, 0, 0, 1, 30, 0x61])); });
    }
  }
  return true;
};
MockSocket.prototype.end = function(data, encoding) {
  if (data) this.write(data, encoding);
  this.emit('end');
  return this;
};
MockSocket.prototype.destroy = function() {
  if (this._destroyed) return this;
  this._destroyed = true;
  this.emit('close', false);
  return this;
};
MockSocket.prototype.setTimeout = function(msecs, callback) {
  if (callback) this.once('timeout', callback);
  return this;
};
MockSocket.prototype.setNoDelay = function() { return this; };
MockSocket.prototype.setKeepAlive = function() { return this; };
MockSocket.prototype.address = function() {
  return { address: this.localAddress, port: this.localPort, family: 'IPv4' };
};

function MockServer(options, connectionCallback) {
  EventEmitter.call(this);
  this._id = genId();
  this.listening = false;
  this._localAddress = undefined;
  this._localPort = undefined;
  this._localFamily = undefined;
  if (typeof options === 'function') {
    this.on('connection', options);
  } else if (typeof connectionCallback === 'function') {
    this.on('connection', connectionCallback);
  }
}
MockServer.prototype = Object.create(EventEmitter.prototype);
MockServer.prototype.constructor = MockServer;
MockServer.prototype.listen = function(options, callback_or_host, callback) {
  var self = this;
  if (typeof options === 'number') {
    var port = options;
    var host = '0.0.0.0';
    var cb;
    if (typeof callback_or_host === 'string') {
      host = callback_or_host;
      cb = callback;
    } else if (typeof callback_or_host === 'function') {
      cb = callback_or_host;
    }
    if (typeof cb === 'function') this.once('listening', cb);
    process.nextTick(function() {
      self._localAddress = host;
      self._localPort = port;
      self._localFamily = 'IPv4';
      self.listening = true;
      self.emit('listening');
    });
  } else if (typeof options === 'object') {
    var listenOptions = { port: options.port || 0, host: options.host || '0.0.0.0' };
    if (typeof callback_or_host === 'function') {
      this.once('listening', callback_or_host);
    }
    process.nextTick(function() {
      self._localAddress = listenOptions.host;
      self._localPort = listenOptions.port;
      self._localFamily = 'IPv4';
      self.listening = true;
      self.emit('listening');
    });
  }
  return this;
};
MockServer.prototype.close = function(callback) {
  if (callback) this.once('close', callback);
  this.listening = false;
  var self = this;
  process.nextTick(function() {
    self.emit('close');
  });
  return this;
};
MockServer.prototype.address = function() {
  if (!this._localAddress) return null;
  return { address: this._localAddress, port: this._localPort, family: this._localFamily };
};
MockServer.prototype.getConnections = function(callback) {
  callback(null, 0);
  return this;
};

function createServer(options, connectionListener) {
  return new MockServer(options, connectionListener);
}

function createConnection(options, callback) {
  var socket = new MockSocket();
  return socket.connect(options, callback);
}

module.exports = {
  Server: MockServer,
  Socket: MockSocket,
  createServer: createServer,
  createConnection: createConnection,
  connect: createConnection,
  isIP: function(input) { return 0; },
  isIPv4: function(input) { return false; },
  isIPv6: function(input) { return false; },
};
